File: Arduino/temperature/python/device.py
#!/usr/bin/env python3
"""
Very simple HTTP server in python for logging requests
Usage::
    ./server.py [<port>]
"""
from http.server import BaseHTTPRequestHandler, HTTPServer
import logging

resp =  b'<root xmlns="urn:schemas-upnp-org:device-1-0" configId="1">\r\n' \
        b'<specVersion>\r\n' \
        b'<major>2</major>\r\n' \
        b'<minor>0</minor>\r\n' \
        b'</specVersion>\r\n' \
        b'<device>\r\n' \
        b'<deviceType>urn:SmartThingsCommunity:device:Temperature:1</deviceType>\r\n' \
        b'<presentationURL>/</presentationURL>\r\n' \
        b'<friendlyName>TemperatureESP8266</friendlyName>\r\n' \
        b'<manufacturer>SmartThingsCommunity</manufacturer>\r\n' \
        b'<manufacturerURL>https://community.smartthings.com</manufacturerURL>\r\n' \
        b'<modelName>LAN Temperature</modelName>\r\n' \
        b'<serialNumber>SN-ESP8266-897</serialNumber>\r\n' \
        b'<UDN>uuid:9487da-SN-ESP8266-897</UDN>\r\n' \
        b'</device></root>\r\n\r\n'
              


class S(BaseHTTPRequestHandler):

    def do_GET(self):
        logging.info("GET request Path: %s\n", str(self.path) )
        
        if self.path == '/TemperatureESP8266.xml':
            logging.info("Sending temperature xml\n")

            #self.send_header("Content-type", "text/xml")
            #self.end_headers()
            self.wfile.write(resp)
            self.send_response(200)

        if self.path == '/refresh?':
            logging.info("Sending refresh\n")

            ref = '{\r\n' \
                  '"status":"HTTP/1.1 200 OK",\r\n' \
                  '"Content-Type":"application/json",\r\n' \
                  '"Cache-Control":"no-cache, private",\r\n' \
                  '"temperature":"18"\r\n' \
                  '}/r/n'

            #self.send_header("Content-type", "text/json")
            #self.end_headers()
            self.wfile.write(bytes(ref, 'utf-8'))
            self.send_response(200)
            

            
    def do_POST(self):
        strPath = str(self.path)
        logging.info("POST request Path: %s\n", strPath)

        
            
def run(server_class=HTTPServer, handler_class=S, port=8191):
    logging.basicConfig(level=logging.INFO)
    server_address = ('0.0.0.0', port)
    httpd = server_class(server_address, handler_class)
    logging.info('Starting httpd...\n')
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()
    logging.info('Stopping httpd...\n')

if __name__ == '__main__':
    from sys import argv

    if len(argv) == 2:
        run(port=int(argv[1]))
    else:
        run()

File: Arduino/temperature/Arduino/Device.cpp
#include <ESP8266WebServer.h>
#include <ESP8266HTTPClient.h>

#include "config.h"

void startPairing();
void loopPair();
void handlePair();

void tempSetup();
float getTemp();

ESP8266WebServer serverDevice(devPort);
String STServer;
String STPort;
String STUUID;
int lastState;

void handleRoot();
void handleNotFound2();
void handleRefresh();
void handlePair();
void handleProfile();
void handlePing();
void LogIt(String msg);

void startDevice() {
  LogIt("Starting device");
  tempSetup();
  pinMode(D7, INPUT_PULLUP);

  STServer  = "";
  STPort    = "";
  STUUID    = "";
  lastState = -1;

  serverDevice.on("/", handleRoot);               // Call the 'handleRoot' function when a client requests URI "/"
  serverDevice.onNotFound(handleNotFound2);        // When a client requests an unknown URI (i.e. something other than "/"), call function "handleNotFound"

  serverDevice.on("/refresh", handleRefresh);
  serverDevice.on("/pair",     handlePair);
  serverDevice.on("/ping",     handlePing);
  
  char prof[1024];
  sprintf(prof, "/%s", devProfile);
  serverDevice.on(prof, handleProfile);
  
  serverDevice.begin();                           // Actually start the server
  Serial.println("HTTP server started");
}

void loopDevice() {
  serverDevice.handleClient();
  loopPair();
  if( digitalRead(D7)!= lastState ) {
    lastState = digitalRead(D7);
    if( STServer != "" ) {
      HTTPClient http;
      WiFiClient client;
      http.begin(client, "http://"+STServer+":"+STPort+"/push-state");
      http.addHeader("Content-Type", "application/json");
      
      int httpResponseCode;
      if( lastState == 1 )
          httpResponseCode = http.POST("{\"uuid\":\"" + STUUID + "\",\n\"contact\":\"closed\"}");
      else
          httpResponseCode = http.POST("{\"uuid\":\"" + STUUID + "\",\n\"contact\":\"open\"}");

    }
  }

}

void handleRefresh() {
  
  char ref[512];
  char contactState[16];
  sprintf(contactState, "closed");
  if( digitalRead(D7)==1 )
      sprintf(contactState, "open");

  sprintf(ref,"{\r\n" \
              "\"status\":\"HTTP/1.1 200 OK\",\r\n" \
              "\"Content-Type\":\"application/json\",\r\n" \
              "\"Cache-Control\":\"no-cache, private\",\r\n" \
              "\"temperature\":\"%f\",\r\n" \
              "\"contact\":\"%s\"\r\n" \
              "}\r\n", getTemp(), contactState);

  LogIt(String(ref));
  serverDevice.send(200, "text/plain", ref);   // Send HTTP status 200 (Ok) and send some text to the browser/client
}

void handlePing() {
  STServer = serverDevice.arg("ip");
  STPort   = serverDevice.arg("port");
  STUUID   = serverDevice.arg("ext_uuid");

  serverDevice.send(200, "text/plain", "Ping");
}

void handlePair() {
  startPairing();
  serverDevice.send(200, "text/plain", "Pairing started");
}

void handleProfile() {
  char resp[1024];
  sprintf(resp, "<root xmlns=\"urn:schemas-upnp-org:device-1-0\" configId=\"1\">\r\n" \
                "<specVersion>\r\n" \
                "<major>2</major>\r\n" \
                "<minor>0</minor>\r\n" \
                "</specVersion>\r\n" \
                "<device>\r\n" \
                "<deviceType>urn:SmartThingsCommunity:device:Light:1</deviceType>\r\n" \
                "<presentationURL>/</presentationURL>\r\n" \
                "<friendlyName>TemperatureESP8266</friendlyName>\r\n" \
                "<manufacturer>SmartThingsCommunity</manufacturer>\r\n" \
                "<manufacturerURL>https://community.smartthings.com</manufacturerURL>\r\n" \
                "<modelName>LAN Temperature</modelName>\r\n" \
                "<serialNumber>SN-ESP8266-899</serialNumber>\r\n" \
                "<UDN>uuid:9487da-SN-ESP8266-899</UDN>\r\n" \
                "</device></root>\r\n\r\n"); 
   serverDevice.send(200, "text/plain", resp);   // Send HTTP status 200 (Ok) and send some text to the browser/client
}

void handleRoot() {
  serverDevice.send(200, "text/plain", "Hello world!");   // Send HTTP status 200 (Ok) and send some text to the browser/client
}

void handleNotFound2(){
  serverDevice.send(404, "text/plain", "404: Not found"); // Send HTTP status 404 (Not Found) when there's no handler for the URI in the request
}